File: Capture.java
void main() {
    // Repeat
    int[] sum = {0};
    repeat(3, i -> {
        sum[0] += i;
    });
    IO.println("sum: " + sum[0]);

    // Gather
    var hub = new Hub();
    gatherHandlers(hub);
    hub.actionPerformed();
}

void repeat(int times, IntConsumer act) {
    for (int i = 0; i < times; ++i) {
        act.accept(i);
    }
}

void gatherHandlers(Hub hub) {
    var items = new ArrayList<Integer>();
    for (int i = 0; i < 3; ++i) {
        int i2 = i;
        hub.onAction(() -> {
            items.add(i2);
            IO.println("items: " + items);
        });
    }
}

class Hub {
    public void onAction(Runnable handler) {
        handlers.add(handler);
    }

    public void actionPerformed() {
        handlers.forEach(Runnable::run);
    }

    List<Runnable> handlers = new ArrayList<>();
}

File: capture.rs
use std::cell::RefCell;
use std::rc::Rc;

fn main() {
    // Repeat
    let mut sum = 0;
    repeat(3, |i| {
        sum += i;
    });
    println!("sum: {}", sum);

    // Gather
    let mut hub = Hub::new();
    init_hub(&mut hub);
    hub.action_performed();
}

fn repeat<F>(times: i32, mut act: F)
where
    F: FnMut(i32),
{
    for i in 0..times {
        act(i);
    }
}

fn init_hub(hub: &mut Hub) {
    let items = Rc::new(RefCell::new(Vec::new()));
    // let mut items = Vec::new();
    for i in 0..3 {
        let items = Rc::clone(&items);
        hub.on_action(Box::new(move || {
            items.borrow_mut().push(i);
            println!("items: {:?}", items.borrow());
            // items.push(i);
            // println!("items: {:?}", items);
        }));
    }
}

struct Hub {
    handlers: Vec<Box<dyn Fn()>>,
}

impl Hub {
    fn new() -> Self {
        Hub {
            handlers: Vec::new(),
        }
    }

    fn on_action(&mut self, handler: Box<dyn Fn()>) {
        self.handlers.push(handler);
    }

    fn action_performed(&self) {
        for handler in &self.handlers {
            handler();
        }
    }
}
